torch/csrc/distributed/c10d/ProcessGroup.cpp

```diff
@@ -115,7 +116,9 @@ ProcessGroup::ProcessGroup(
 ProcessGroup::ProcessGroup(int rank, int size)
     : rank_(rank), size_(size), backendType_(BackendType::UNDEFINED) {}
 
-ProcessGroup::~ProcessGroup() = default;
+ProcessGroup::~ProcessGroup() {
+  c10d::unregister_completed_works();
+}
```

Comment:
Unfortunately there is no reliable callback method where we can insert the "unregister_work()" cleanup call for the pure-eager case (`Work::finish()` is only called in Gloo PG but not NCCL PG, thus we can't rely on it). Instead we can do `unregister_completed_works()` here to do the cleanup for completed works, before throwing the warning in `~WorkRegistry()` for the uncompleted works. `ProcessGroup` instance bound to Python should always be destroyed by Python shutdown sequence before the C++ per-thread static object `WorkRegistry`, so doing the cleanup here should work.